contracts/farm/LiquidityMining.sol

```diff
@@ -0,0 +1,567 @@
+		//update user accumulated reward
+		if (user.amount > 0) {
+			//add reward for the previous amount of deposited tokens
+			uint256 accumulatedReward = user.amount.mul(pool.accumulatedRewardPerShare).div(PRECISION).sub(user.rewardDebt);
+			user.accumulatedReward = user.accumulatedReward.add(accumulatedReward);
+		}
+	}
+
+	function _updateRewardDebt(PoolInfo storage pool, UserInfo storage user) internal {
+		//reward accumulated before amount update (should be subtracted during next reward calculation)
+		user.rewardDebt = user.amount.mul(pool.accumulatedRewardPerShare).div(PRECISION);
+	}
+
+	/**
+	 * @notice Send reward in SOV to the lockedSOV vault.
+	 * @param _user The user info, to get its reward share.
+	 * @param _userAddress The address of the user, to send SOV in its behalf.
+	 */
+	function _transferReward(
+		UserInfo storage _user,
+		address _userAddress,
+		bool _isClaimingReward
+	) internal {
+		uint256 userAccumulatedReward = _user.accumulatedReward;
+
+		/// @dev Transfer if enough SOV balance on this LM contract.
+		uint256 balance = SOV.balanceOf(address(this));
+		if (balance >= userAccumulatedReward) {
+			totalUsersBalance = totalUsersBalance.sub(userAccumulatedReward);
+			_user.accumulatedReward = 0;
+
+			/// @dev Instead of transferring the reward to the LP (user),
+			///   deposit it into lockedSOV vault contract, but first
+			///   SOV deposit must be approved to move the SOV tokens
+			///   from this LM contract into the lockedSOV vault.
+			SOV.approve(address(lockedSOV), userAccumulatedReward);
+			lockedSOV.deposit(_userAddress, userAccumulatedReward, basisPoint);
+
+			if (_isClaimingReward) {
+				lockedSOV.withdrawAndStakeTokensFrom(_userAddress);
+			}
+
+			/// @dev Event log.
+			emit RewardClaimed(_userAddress, userAccumulatedReward);
+		}
+	}
+
+	/**
+	 * @notice withdraws pool tokens without transferring reward tokens
+	 * @param _poolToken the address of pool token
+	 * @dev EMERGENCY ONLY
+	 */
+	function emergencyWithdraw(address _poolToken) public {
+		uint256 poolId = _getPoolId(_poolToken);
+		PoolInfo storage pool = poolInfoList[poolId];
+		UserInfo storage user = userInfoMap[poolId][msg.sender];
+
+		uint256 userAmount = user.amount;
+		user.amount = 0;
+		user.rewardDebt = 0;
+		user.accumulatedReward = 0;
+		pool.poolToken.safeTransfer(address(msg.sender), userAmount);
+		emit EmergencyWithdraw(msg.sender, _poolToken, userAmount);
+	}
+
+	/**
+	 * @notice returns pool id
+	 * @param _poolToken the address of pool token
+	 */
+	function getPoolId(address _poolToken) public view returns (uint256) {
+		return _getPoolId(_poolToken);
+	}
+
+	function _getPoolId(address _poolToken) internal view returns (uint256) {
+		uint256 poolId = poolIdList[_poolToken];
+		require(poolId > 0, "Pool token not found");
+		return poolId - 1;
+	}
+
+	/**
+	 * @notice returns count of pool tokens
+	 */
+	function getPoolLength() external view returns (uint256) {
+		return poolInfoList.length;
```

Comment:
If the index starts at 0, shouldn't this be length - 1?

Reply:
then it would return length 4 in case there are 5 entries. 
this is not supposed to return the last index, but the length of the array.

Reply:
My bad, this is consistent.